Python/pywarpx/picmi.py

```diff
@@ -555,6 +555,62 @@ def initialize_inputs(self, laser):
             laser.laser.profile_t_peak = (self.position[2] - laser.centroid_position[2])/constants.c  # The time at which the laser reaches its peak (in seconds)
 
 
+class ConstantAppliedField(picmistandard.PICMI_ConstantAppliedField):
+    def initialize_inputs(self):
+        # Note that lower and upper_bound are not used by WarpX
+
+        if (self.Ex is not None or
+            self.Ey is not None or
+            self.Ez is not None):
+            pywarpx.particles.E_ext_particle_init_style = 'constant'
+            pywarpx.particles.E_external_particle = [self.Ex or 0., self.Ey or 0., self.Ez or 0.]
+
+        if (self.Bx is not None or
+            self.By is not None or
+            self.Bz is not None):
+            pywarpx.particles.B_ext_particle_init_style = 'constant'
+            pywarpx.particles.B_external_particle = [self.Bx or 0., self.By or 0., self.Bz or 0.]
+
+
+class AnalyticAppliedField(picmistandard.PICMI_AnalyticAppliedField):
+    def initialize_inputs(self):
+        # Note that lower and upper_bound are not used by WarpX
+        for k,v in self.user_defined_kw.items():
+            setattr(pywarpx.my_constants, k, v)
+
+        if (self.Ex_expression is not None or
+            self.Ey_expression is not None or
+            self.Ez_expression is not None):
+            pywarpx.particles.E_ext_particle_init_style = 'parse_e_ext_particle_function'
+            pywarpx.particles.__setattr__('Ex_external_particle_function(x,y,z,t)', self.Ex_expression)
+            pywarpx.particles.__setattr__('Ey_external_particle_function(x,y,z,t)', self.Ey_expression)
+            pywarpx.particles.__setattr__('Ez_external_particle_function(x,y,z,t)', self.Ez_expression)
+
+        if (self.Bx_expression is not None or
+            self.By_expression is not None or
+            self.Bz_expression is not None):
+            pywarpx.particles.B_ext_particle_init_style = 'parse_b_ext_particle_function'
+            pywarpx.particles.__setattr__('Bx_external_particle_function(x,y,z,t)', self.Bx_expression)
+            pywarpx.particles.__setattr__('By_external_particle_function(x,y,z,t)', self.By_expression)
+            pywarpx.particles.__setattr__('Bz_external_particle_function(x,y,z,t)', self.Bz_expression)
+
+
+class Mirror(picmistandard.PICMI_Mirror):
+    def initialize_inputs(self):
+        try:
+            pywarpx.warpx.num_mirrors
+        except AttributeError:
+            pywarpx.warpx.num_mirrors = 0
+            pywarpx.warpx.mirror_z = []
+            pywarpx.warpx.mirror_z_width = []
+            pywarpx.warpx.mirror_z_npoints = []
```

Comment:
```suggestion
            pywarpx.warpx.num_mirrors = 0
            pywarpx.warpx.mirror_z = []
            pywarpx.warpx.mirror_z_width = []
            pywarpx.warpx.mirror_z_npoints = []
```
This is just a minor question/suggestion.
I think that these are the defaults already on WarpX.
I would say that if the user doesn't write explicitly `pywarpx.warpx.num_mirrors = 0`, then there is no need to do this initialization in PICMI.
Or is it incompatible with PICMI defaults?

Reply:
This is creating the input variables in the picmi implementation that will be filled by the user's value below. Without this code, those Python variables would not be created anywhere. If a Mirror object is never used, those variables, num_mirrors etc, would not be created and would not appear in the input for WarpX. Does that help?

Reply:
Yes, thank you @dpgrote !